wal/wal.go

```diff
@@ -899,6 +884,232 @@ func (r *Reader) Offset() int64 {
+		temp, n, err := readRecord(r.buf[r.readIndex:r.writeIndex], r.hdr[:], r.total)
+		r.readIndex += n
+		r.total += int64(n)
+		if err != nil {
+			r.err = err
+			return false
+		}
+		if temp == nil {
+			return false
+		}
+		fullRec, err := validateRecord(recType(r.hdr[0]), r.index)
+		if err != nil {
+			r.err = err
+			r.index = 0
+			return false
+		}
+
+		r.rec = append(r.rec, temp...)
+		if fullRec {
+			r.index = 0
+			return true
+		}
+
+		// Only increment i for non-zero records since we use it
+		// to determine valid content record sequences.
+		r.index++
+	}
+}
+
+// Returns true if we have a full record, false if it's partial. Returns an
+// error if the full or partial record is invalid based on i, true otherwise.
+func validateRecord(typ recType, i int) (bool, error) {
+	switch typ {
+	case recFull:
+		if i != 0 {
+			return false, errors.New("unexpected full record")
+		}
+		return true, nil
+	case recFirst:
+		if i != 0 {
+			return false, errors.New("unexpected first record, dropping buffer")
+		}
+		return false, nil
+	case recMiddle:
+		if i == 0 {
+			return false, errors.New("unexpected middle record, dropping buffer")
+		}
+		return false, nil
+	case recLast:
+		if i == 0 {
+			return false, errors.New("unexpected last record, dropping buffer")
+		}
+		return true, nil
+	default:
+		return false, errors.Errorf("unexpected record type %d", typ)
+	}
+}
+
+// Read a sub-record (see recType) from the buffer. It could potentially
+// be a full record (recFull) if the record fits within the bounds of a single page.
+// Returns a byte slice of the record data read, the number of bytes read, and an error
+// if there's a non-zero byte in a pager term record or the record checksum fails.
+func readRecord(buf []byte, header []byte, total int64) ([]byte, int, error) {
+	readIndex := 0
+	header[0] = buf[0]
+	readIndex++
+	total++
+
+	// This rest of this function is mostly from Reader.Next
+	typ := recType(header[0])
+	// Gobble up zero bytes.
+	if typ == recPageTerm {
+		// We are pedantic and check whether the zeros are actually up to a page boundary.
+		// It's not strictly necessary but may catch sketchy state early.
+		k := pageSize - (total % pageSize)
+		if k == pageSize {
+			return nil, 1, nil // Initial 0 byte was last page byte.
+		}
+
+		if k <= int64(len(buf)-readIndex) {
+			temp := make([]byte, k)
+			copied := copy(temp, buf[readIndex:int64(readIndex)+k])
+			readIndex += copied
+			for _, v := range temp {
+				if v != 0 {
+					return nil, readIndex, errors.New("unexpected non-zero byte in page term bytes")
+				}
+			}
+			return nil, readIndex, nil
+		}
+		// Not enough bytes to read the rest of the page term rec.
+		// This theoretically should never happen, since we're now always reading a page at a time
+		// and then processing sub-records from that page.
+		// Treat this the same as an EOF, it's an error we would expect to see.
+		return nil, 0, io.EOF
+	}
+
+	// we probably don't need this anymore
```

Comment:
Comments should be capitalised and end with period.

Also, why won't we need this check?

Reply:
We do. I added this comment before I understood the page flush logic.